### DISCUSSION TITLE
Feed external data to watchface

### DISCUSSION THREAD

**Discussion by kukuchta** [zepp-health/zeppos-forum#330] [Q&A]
Hello! 
I'm looking for a way to fetch some data in Zepp OS 3.0 from external service through Fetch API to Side Service in the Zepp App and then send it through BLE channel to the watch. 

The missing part is - how to send this data not to the Device App, but directly to the Watchface App? 

Or maybe there is some way to communicate between Device App and a Watchface?

Purpose of this would be displaying custom realtime data on the watchface regarding blood glucose level (essential for people with t1 diabetes).

I would appreciate any information.

**Comment by silver-zepp** [comment]
Hey, the watch faces don't support Side Service, therefore this way of communication is not possible. You have to rely on a mini app. 
Here's a solution you are looking for: [AutoGUI-3.0-wf-app-comms-example](https://github.com/silver-zepp/zeppos-autogui/tree/master/other-autogui-examples/zeppos-3.0-wf-app-comms)

**Reply by oakeley** [reply]
@silver-zepp the mini app method used to work for us in v1 and v2 but changes to the permissions in v3 between mini app and watchface make the last step very difficult. In addition the extreme power saving causes the mini app to fall asleep in v3 so it doesn't work reliably for a background process because the watchface can't wake it up (imagine you just look at your watch while driving, the data will not be current if the mini app is asleep).
Do you have a command that a watchface can send to wake up or restart the mini app background collection process?
I tried installing your GlucoCore mini-app and GlucoCore watchface but the screen is only blank and it doesn't have an obvious line commented out for the phone URL / Nightscout URL from which glucose can be pulled (the screen is just black on the Balance)

**Reply by oakeley** [reply]
@silver-zepp can you make an update to your GlucoCore miniapp and watchface that will be able to pull the data from the phone  URL:
http://localhost:29863/info.json
(This is the standard for the entire community)... 

The format is for the page content from the URL is like this:

{"bg":{"delta":"+0.0","isHigh":false,"isLow":false,"isStale":false,"time":**1713000682601**,"trend":"Flat","val":"**7.0**"},"external":{},"pump":{"bat":0.0,"iob":0.0,"reservoir":0.0},"status":{"bat":91,"isMgdl":false,"now":1713000719017},"treatment":{}}

Simply displaying would be fine or for bonus points extract the "val" which for me is "7.0" (mmol/) at the moment. The "time" is in Unix time and "ideally" would not be more than 8-10 minutes in the past from the current "watch time" (if it is then drawing a line through the number to indicate that it is stale would be appropriate). You could also use the "isStale" flag for this purpose

**Reply by oakeley** [reply]
@silver-zepp a challenge for the future is that the URL only contains "the current" glucose value... most people want to have the last few hours displayed as a graph. There are various hacks that can work to get these data uploaded but in principle for an Abbott CGM the phone gets data every minute and for Dexcom and the rest the phone gets data every 4' can you capture the glucose "val" from the URL every minute in the background and write that to file (that stores 24 hours of data) to make a glucose graph? Will this have a huge battery hit? 
Definitely priority "20" after the "read the glucose from the JSON URL" but ultimately folks in the community will ask for it

**Reply by kukuchta** [reply]
I can only add that there is a huge community of amazfit watches users that buy them solely for this purpose, to trace their diabetes parameters. Without this feature working reliable, there is no point going for newest models. 

I also just recently bought the GTR4, happily updated it to zepp OS 3.0 only to find that now it is completely unusable for my purpose.

**Reply by oakeley** [reply]
@kukuchta exactly. It was a huge disappointment that this feature was broken in v3.0
With luck @silver-zepp's solution will eventually make it work again.
Good for them (sell more watches)... Good for us (don't get random updates from v2 to v3 that break the watches completely and risk our lives)

**Reply by kukuchta** [reply]
@silver-zepp Could you please comment on the issues with permissions and sleeping raised by @oakeley? Without those things resolved, our community would be tied to older watches and forced to avoid any OS upgrades beyond 2.x.

**Comment by oakeley** [comment]
I guess adding something like:
  const url = 'http://localhost:29863/info.json'
  const { body: { data = {} } = {} } = await fetch(url)
to the "getRandomGluco()" function would be a start but the fact that the random glucose value is not displayed on the watchface is a worry...
@silver-zepp looking forward to testing the next update. Thank you for your support

**Comment by jiresongit** [comment]
This would be fantastic if it could be made to work on zepp3... From another diabetic.

**Comment by oakeley** [comment]
> I can only add that there is a huge community of amazfit watches users that buy them solely for this purpose, to trace their diabetes parameters. Without this feature working reliable, there is no point going for newest models.
> 
> I also just recently bought the GTR4, happily updated it to zepp OS 3.0 only to find that now it is completely unusable for my purpose.

You could probably try to return your GTR4 and ask for a replacement GTR4 and then be very careful to NOT click "update". A bit unfair on the shop you bought it from but better than having to buy an earlier watch

**Reply by kukuchta** [reply]
Yeah, a little unfair to me, and still, returning anything back to China once it arrived safely is a long path to go😉

I believe this issue could be solved here in a civilized way. Zepp OS development wouldn stop at this version anyway, there will be new watches and new updates.

**Comment by ShukriChiu** [comment]
please add this feature

**Comment by hsjg53i** [comment]
there are a lot of us that would like for this feature to be added

**Comment by Chriz76** [comment]
> Hey, the watch faces don't support Side Service, therefore this way of communication is not possible. You have to rely on a mini app. Here's a solution you are looking for: [AutoGUI-3.0-wf-app-comms-example](https://github.com/silver-zepp/zeppos-autogui/tree/master/other-autogui-examples/zeppos-3.0-wf-app-comms)

@oakeley and @silver-zepp this code looks like the watchface supports communicating with a side service. There is even a comment about zepp os 3: https://github.com/bigdigital/zeppos_watchdrip_timer_wf/blob/master/utils/watchdrip/watchdrip.js

So it should be working, doesn't it?

**Reply by mrklos** [reply]
Hello! Any update on this topic?
